Handle race conditions from deleting a message and adding/removing related fields simultaneously.

abhijeetbodas2001 opened the pull request:
## [Do not merge]

This replaces #16721.

abhijeetbodas2001 commented:
This PR now uses `captureOnCommitCallbacks`, but there's an issue with database transactions which I found while testing this.
https://chat.zulip.org/#narrow/stream/3-backend/topic/message.20edit.20race.20.2316502/near/1184676

abhijeetbodas2001 commented:
I've tested all scenarios I could think of, and this is ready for review.
@mateuszmandera could you take a look?

timabbott commented:
I merged the first commit as 8bcdbc720e3a7203fc0056bb8c81d5847b39464b, just to head off merge conflicts.

timabbott commented on zerver/lib/test_classes.py:
```diff
@@ -1266,7 +1266,10 @@ def tornado_redirected_to_list(self, lst: List[Mapping[str, Any]]) -> Iterator[N
-        yield
+
+        with self.captureOnCommitCallbacks(execute=True):
+            yield
```
I'd put a polished version commit message explanation in a nice block comment here, since it's important to understanding what's happening here in our test infrastructure.  

(Also the existing comments for this function seem misplaced; I'd expect them to have been above the hacky functions).

timabbott replied:
As a sidenote, one of the most important parts of having a readable codebase is asking the question "Where does this explanation belong?" -- a comment, a commit message, a function name, documentation in ReadTheDocs with a link are all valid answers, and choosing the right one is a big part of making a codebase happy.

abhijeetbodas2001 replied:
Done.

timabbott commented on zerver/tests/test_events.py:
```diff
@@ -275,7 +275,10 @@ def verify_action(
         )
-        action()
+
+        with self.captureOnCommitCallbacks(execute=True):
```
Here I'd add a brief comment referencing the main comment `See ZulipTestCase.tornado_redirected_to_list for details`

timabbott commented on zerver/lib/message.py:
```diff
@@ -655,7 +655,10 @@ def build_dict_from_raw_db_row(row: RawReactionRow) -> Dict[str, Any]:
+    user_profile: UserProfile,
+    message_id: int,
+    lock_message: bool = False,
+    lock_usermessage: bool = False,
```
These probably deserve a bit of documentation in the docstring about when to use them.

abhijeetbodas2001 replied:
Done.

timabbott commented on zerver/views/message_edit.py:
```diff
@@ -143,13 +143,14 @@ def validate_can_delete_message(user_profile: UserProfile, message: Message) ->
+    message, ignored_user_message = access_message(user_profile, message_id, lock_message=True)
     validate_can_delete_message(user_profile, message)
     try:
         do_delete_messages(user_profile.realm, [message])
```
What will happen if someone had locked a UserMessage row when we call this?  I guess what I'd hope for is that we just finish the other transaction happily and then run this.  (In which case it should be illegal to lock a UserMessage if you're later going to want to lock the Message within that transaction).

One possible theory for not needing to think about this sort of question is that we should just always lock the message object (and never lock UserMessage objects).

timabbott commented on zerver/lib/push_notifications.py:
```diff
@@ -856,47 +856,53 @@ def handle_push_notification(user_profile_id: int, missed_message: Dict[str, Any
+        try:
+            (message, user_message) = access_message(
+                user_profile, missed_message["message_id"], lock_usermessage=True
+            )
```
Similarly, I think we could explain that we lock the UserMessage because ...

timabbott commented on zerver/tests/test_submessage.py:
```diff
@@ -115,7 +115,8 @@ def test_endpoint_success(self) -> None:
         with mock.patch("zerver.lib.actions.send_event") as m:
-            result = self.client_post("/json/submessage", payload)
+            with self.captureOnCommitCallbacks(execute=True):
+                result = self.client_post("/json/submessage", payload)
```
Can we instead have a prep commit refactor this test to use `tornado_redirected_to_list`?  It'd be nice to contain `self.captureOnCommitCallbacks` calls to just the two place we use this.

abhijeetbodas2001 replied:
This was done in #18626.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2161,7 +2165,7 @@ def do_add_submessage(
     target_user_ids = [um.user_profile_id for um in ums]
 
-    send_event(realm, event, target_user_ids)
+    transaction.on_commit(lambda: send_event(realm, event, target_user_ids))
```
I've sorta forgotten the background here, but is the failure mode obvious if we fail to move the `send_event` call into `transaction.on_commit()`?  (E.g. tests fail).  If not, we may want to think about how to ensure that's the case.

abhijeetbodas2001 replied:
No, if someone were to move `send_event` outside `on_commit` later on, none of the test would fail.
I'm not sure how we can test this though 🤔 Maybe mocking `send_event` to throw an error?

abhijeetbodas2001 replied:
Update: Mocking `send_event` is actually useful (https://github.com/zulip/zulip/pull/18559#issuecomment-854667223)

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2226,20 +2226,20 @@ def do_add_reaction(
-        # This can happen when a race results in the check in views
-        # code not catching an attempt to double-add a reaction, or
-        # perhaps if the emoji_name/emoji_code mapping is busted.
-        raise JsonableError(_("Reaction already exists."))
```
Can you tweak the commit message to explain why we get to delete this a bit more explicitly?  I think the story is "... because `lock_message` will wait until the duplicate transaction finishes".

"""
As a nice side-effect, we also eliminate the previous race condition from double
adding reactions (by waiting for the message row lock), and thus get to delete 
somewhat messy `IntegrityError` error handling logic.
"""

abhijeetbodas2001 replied:
Done.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -2226,20 +2226,20 @@ def do_add_reaction(
+    """Should be called while holding a SELECT FOR UPDATE lock
+    (e.g. via access_message(..., lock_message=True)) on the
+    Message row, to prevent race conditions.
+    """
```
Thinking more about these comments, could we replace these with something like (pseudocode):

`assert message.row_is_locked`

so that the code is self-explanatory.  And we could potentially just set that property in `access_message` in the initial commit that adds support for `lock_message`.  (And have it be a default value that's false on the `Message` class, so one doesn't need to `getattr`?)

abhijeetbodas2001 replied:
Update: Mocking `send_event` is actually useful (https://github.com/zulip/zulip/pull/18559#issuecomment-854667223)

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5475,17 +5475,17 @@ def do_update_message_flags(
+            UserMessage.objects.update_or_create(
+                user_profile=user_profile,
+                message=message,
+                flags=UserMessage.flags.historical | UserMessage.flags.read,
```
Is the right way to call `update_or_create`?  I think you need to rearrange how `flags` are passed for it, but am not sure.

abhijeetbodas2001 replied:
The `update` case doesn't really do anything here (other than avoiding `IntegrityErrors` from duplicate entries), because if the reaction code had created a UserMessage in the meanwhile, it would be the same as what this codepath would have created.

abhijeetbodas2001 replied:
Update: I have made this code more explicit by doing a `.exists()` check, and also adding a comment to explain in what situations there may be a `UserMessage` already present.

timabbott commented:
OK, I'm feeling pretty good about this PR.  A few ideas:
* Can we drop the `lock_usermessage` stuff for now, and leave that for a follow-up PR?  I feel like it has various tricky things to think about, and it might be worth getting all the `lock_message` stuff in before turning to that in full.
* I note this detail in the [Django documentation](https://docs.djangoproject.com/en/3.2/ref/models/querysets/#select-for-update): "By default, select_for_update() locks all rows that are selected by the query. For example, rows of related objects specified in select_related() are locked in addition to rows of the queryset’s model. If this isn’t desired, specify the related objects you want to lock in select_for_update(of=(...)) using the same fields syntax as select_related(). Use the value 'self' to refer to the queryset’s model.".  I think we probably do want our `lock_message` logic to pass `of=(self)`, i.e. not try to lock foreign keys, since only the Message object itself has these races; we're not super worried about the UserProfile object, and want to take minimal locks.

mateuszmandera commented on zerver/views/message_edit.py:
```diff
@@ -143,13 +143,14 @@ def validate_can_delete_message(user_profile: UserProfile, message: Message) ->
 
+@transaction.atomic
 @has_request_variables
 def delete_message_backend(
```
This isn't the only relevant codepath - e.g. there's also the "delete topic" API where I'm pretty sure we'll want to lock too. `git grep do_delete_messages` should give all the places we need to take a look at.

abhijeetbodas2001 replied:
Updated to handle the `delete topic` codepath too.

mateuszmandera commented:
@timabbott There is also the case of scheduled archiving (the retention.py system) that hasn't been addressed yet - that is, the archiving cron job while it's doing its things will have various races with message deletion/editing etc. Perhaps we can leave that for a follow-up PR, but worth noting at least.

timabbott commented:
Yes, agreed, I think that's fine to leave as a follow-up project for now but let's not lose track of it.

(Probably we'll want to open a few issues when we close this for the known follow-ups).

timabbott commented:
@abhijeetbodas2001 let me know when this is ready for a next review.

abhijeetbodas2001 commented:
I've addressed all review comments except https://github.com/zulip/zulip/pull/18559#discussion_r638386127.

Changes are:
- Dropped the `UserMessage` changes and used `.select_for_update(of=["self"])` as suggested in https://github.com/zulip/zulip/pull/18559#issuecomment-847459652
- Added comments wherever suggested.
- Managed to write failing tests for the `on_commit` hooks of the form:
  ```python
    with self.tornado_redirected_to_list([], expected_num_events=1):
            with mock.patch("zerver.lib.actions.send_event") as m:
                m.side_effect = AssertionError()
                action()
  ```
  These work because of scoping -- `send_event` calls will be run because of `captureOnCommitCallbacks` only *after* control has left the `captureOnCommitCallbacks` context manager, but by then, our mock has done what we wanted it to 🙂 
- Made changes to lock messages even when deleting a topic. This is logic independent of `access_message`

abhijeetbodas2001 commented:
Re https://github.com/zulip/zulip/pull/18559#discussion_r638386127, I'm not sure if I understand what needs to be done. Do you mean we create new field on `Message` class for this purpose? I did not understand how that would help. A hypothetical bad caller which ignores the docstring could still set that field as `true`, right?

Another idea is that, there's this field used by Django internally -> [`QuerySet.query.select_for_update: bool`](https://github.com/django/django/blob/ca9872905559026af82000e46cde6f7dedc897b6/django/db/models/query.py#L1052). If we refactor `access_message` to return a `QuerySet` (instead of a `Message`) we could probably use that directly to assert if the row is locked.

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5532,17 +5532,24 @@ def do_update_message_flags(
-            message=message,
-            flags=UserMessage.flags.historical | UserMessage.flags.read,
-        )
+        with transaction.atomic():
```
Could the `transaction.atomic()` be around a larger block, such that we don't need to do a second UserMessage check?

abhijeetbodas2001 replied:
I don't think so. The problem really is that we lock the message *after* we realize there's no UserMessage present.

abhijeetbodas2001 replied:
I adjusted the code to avoid the new query -- `access_message` already queries for `UserMessage` anyways 😅

timabbott commented on zerver/lib/actions.py:
```diff
@@ -5532,17 +5532,24 @@ def do_update_message_flags(
         if flag != "starred":
             raise JsonableError(_("Invalid message(s)"))
         # Validate that the user could have read the relevant message
```
This comment belongs inside the `transaction.atomic`.

abhijeetbodas2001 replied:
Done

alexmv commented:
https://sentry.io/share/issue/775a4e4173314d34a2e75751e02ed932/ is a race of `embed_links` with message deletion, which I don't think this covered by this yet.

mateuszmandera commented:
I think we just want to handle the DoesNotExist case + lock the Message row so that it doesn't get deleted while being processed by the queue worker?

Something like 
```diff
diff --git a/zerver/worker/queue_processors.py b/zerver/worker/queue_processors.py
index d94237905e..d72132c452 100644
--- a/zerver/worker/queue_processors.py
+++ b/zerver/worker/queue_processors.py
@@ -738,25 +738,31 @@ class FetchLinksEmbedData(QueueProcessingWorker):
                 "Time spent on get_link_embed_data for %s: %s", url, time.time() - start_time
             )
 
-        message = Message.objects.get(id=event["message_id"])
-        # If the message changed, we will run this task after updating the message
-        # in zerver.lib.actions.check_update_message
-        if message.content != event["message_content"]:
-            return
-        if message.content is not None:
-            query = UserMessage.objects.filter(
-                message=message.id,
-            )
-            message_user_ids = set(query.values_list("user_profile_id", flat=True))
+        with transaction.atomic():
+            try:
+                message = Message.objects.select_for_update().get(id=event["message_id"])
+            except Message.DoesNotExist:
+                logging.info("Message %s no longer exists.", event["message_id"])
+                return
 
-            # Fetch the realm whose settings we're using for rendering
-            realm = Realm.objects.get(id=event["message_realm_id"])
+            # If the message changed, we will run this task after updating the message
+            # in zerver.lib.actions.check_update_message
+            if message.content != event["message_content"]:
+                return
+            if message.content is not None:
+                query = UserMessage.objects.filter(
+                    message=message.id,
+                )
+                message_user_ids = set(query.values_list("user_profile_id", flat=True))
 
-            # If rendering fails, the called code will raise a JsonableError.
-            rendered_content = render_incoming_message(
-                message, message.content, message_user_ids, realm
-            )
-            do_update_embedded_data(message.sender, message, message.content, rendered_content)
+                # Fetch the realm whose settings we're using for rendering
+                realm = Realm.objects.get(id=event["message_realm_id"])
+
+                # If rendering fails, the called code will raise a JsonableError.
+                rendered_content = render_incoming_message(
+                    message, message.content, message_user_ids, realm
+                )
+                do_update_embedded_data(message.sender, message, message.content, rendered_content)
 
 
 @assign_queue("outgoing_webhooks")

```

mateuszmandera commented:
Also, looking at `do_update_embedded_data` it seems to have a redundant `atomic()` decorator  from 5 years ago?
```
# We use transaction.atomic to support select_for_update in the attachment codepath.
@transaction.atomic
def do_update_embedded_data(
```

I don't see an attachment codepath here, so we can probably just remove the decorator?   And even in the case where we want to keep it, it's doing `send_event(user_profile.realm, event, list(map(user_info, ums)))` inside it, which is a pattern we want to avoid

timabbott commented:
We may want to avoid locking the `Message` row while calling `render_incoming_message`, just because that's a potentially slow operation and every situation where we hold a lock for a long time is costly.

One idea would be to have a shorter atomic block with locking just for the end bit in `do_update_embedded_data`.

> Also, looking at do_update_embedded_data it seems to have a redundant atomic() decorator from 5 years ago?

Hmm, I don't know the background on that, but that does seem worth changing.

Do we also want to change all use of transaction.atomic to use `savepoint=False` across the codebase, and lint for it, so that we don't have to consider the possibility of nested `@transaction.atomic`?

abhijeetbodas2001 commented:
I think `durable` (new in Django 3.2) might be better for ensuring no nested `transaction.atomic`s. From Django's docs-
> It is sometimes useful to ensure an atomic block is always the outermost atomic block, ensuring that any database changes are committed when the block is exited without errors. This is known as durability and can be achieved by setting durable=True. If the atomic block is nested within another it raises a RuntimeError.

And then maybe we could have a linter rule to enforce `durable=True` is always passed in `transaction.atomic`?

timabbott commented:
Yeah, `durable=True` works for places where we intend that -- I think we might have a place where we want `savepoint=False` because the function can be called from multiple code paths, only some of which already have a transaction.

I'm fine with a lint rule to enforce that we pass one of those two parameters.  (Though maybe that'll fail tests?  Not sure how the testing transactions work).

timabbott commented:
Just an update that the parts of this that are already merged are now now running happily in production.
